spring-data-mongodb/src/main/java/org/springframework/data/mongodb/SessionAwareMethodInterceptor.java

```diff
@@ -0,0 +1,211 @@
+	 * @param databaseDecorator a {@link ClientSessionOperator} used to create the proxy for an imperative / reactive
+	 *          {@code MongoDatabase}.
+	 * @param collectionType the MongoDB collection type.
+	 * @param collectionDecorator a {@link ClientSessionOperator} used to create the proxy for an imperative / reactive
+	 *          {@code MongoCollection}.
+	 * @param <T> target object type.
+	 */
+	public <T> SessionAwareMethodInterceptor(ClientSession session, T target, Class<D> databaseType,
+			ClientSessionOperator<D> databaseDecorator, Class<C> collectionType,
+			ClientSessionOperator<C> collectionDecorator) {
+
+		Assert.notNull(session, "ClientSession must not be null!");
+		Assert.notNull(target, "Target must not be null!");
+		Assert.notNull(databaseType, "Database type must not be null!");
+		Assert.notNull(databaseDecorator, "Database ClientSessionOperator must not be null!");
+		Assert.notNull(collectionType, "Collection type must not be null!");
+		Assert.notNull(collectionDecorator, "Collection ClientSessionOperator must not be null!");
+
+		this.session = session;
+		this.target = target;
+		this.databaseType = ClassUtils.getUserClass(databaseType);
+		this.collectionType = ClassUtils.getUserClass(collectionType);
+		this.collectionDecorator = collectionDecorator;
+		this.databaseDecorator = databaseDecorator;
+
+		this.targetType = ClassUtils.isAssignable(databaseType, target.getClass()) ? databaseType : collectionType;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see org.aopalliance.intercept.MethodInterceptor(org.aopalliance.intercept.MethodInvocation)
+	 */
+	@Nullable
+	@Override
+	public Object invoke(MethodInvocation methodInvocation) throws Throwable {
+
+		if (requiresDecoration(methodInvocation.getMethod())) {
+
+			Object target = methodInvocation.proceed();
+			if (target instanceof Proxy) {
+				return target;
+			}
+
+			return decorate(target);
+		}
+
+		if (!requiresSession(methodInvocation.getMethod())) {
+			return methodInvocation.proceed();
+		}
+
+		Optional<Method> targetMethod = METHOD_CACHE.lookup(methodInvocation.getMethod(), targetType);
+
+		return !targetMethod.isPresent() ? methodInvocation.proceed()
+				: ReflectionUtils.invokeMethod(targetMethod.get(), target,
+						prependSessionToArguments(session, methodInvocation));
+	}
+
+	private boolean requiresDecoration(Method method) {
+
+		return ClassUtils.isAssignable(databaseType, method.getReturnType())
+				|| ClassUtils.isAssignable(collectionType, method.getReturnType());
+	}
+
+	@SuppressWarnings("unchecked")
+	protected Object decorate(Object target) {
+
+		return ClassUtils.isAssignable(databaseType, target.getClass()) ? databaseDecorator.apply(session, target)
+				: collectionDecorator.apply(session, target);
+	}
+
+	private static boolean requiresSession(Method method) {
+
+		if (method.getParameterCount() == 0
+				|| !ClassUtils.isAssignable(ClientSession.class, method.getParameterTypes()[0])) {
+			return true;
+		}
+
+		return false;
```

Comment:
just an observation. this can be simplified to :
` return method.getParameterCount() == 0
                || !ClassUtils.isAssignable(ClientSession.class, method.getParameterTypes()[0]);`

Reply:
Thanks a lot for having a look. We leave some conditionals in their extended form to simplify readability. Eliminating the `if` increases density and it's harder to comprehend what's going on. The extended form with blank lines help to separate the actual cases.

Reply:
ok, got it, thank!